forem/forem
# Pusher Beams for Mobile Push Notifications refactor

## What type of PR is this? (check all applicable)


## Description

Looking forward to the Forem ecosystem, we want a way for consumers of each Forem Instance to be able to register for Push Notifications.

It doesn't matter multiple apps coexist, they can all register for Push Notification delivery and the Forem instance (Rails server) will be able to distribute those PNs to all clients.

**TODO:**
- [x] Refactor Pusher Beams to use `SiteConfig` configurable in `/admin/config`
- [x] Add the push notifications instance ID to the DOM in the Forem meta tags
- [ ] Add namespace for multiple forems
- [ ] Create endpoints that will support [Authenticated Users](https://pusher.com/docs/beams/concepts/authenticated-users)

## Related Tickets & Documents

Closes https://github.com/forem/InternalProjectPlanning/issues/263

## QA Instructions, Screenshots, Recordings

You can checkout `/admin/config` to make sure the SiteConfig works as expected

![Screen Shot 2020-12-23 at 08 12 49](https://user-images.githubusercontent.com/6045239/103003643-bee11e00-44f6-11eb-8bb4-b2abcc2c85bf.png)

I feel like a broken record, but this is not easy to QA. You'll need to enable Pusher Beams locally and register a mobile app, then send PNs to that device.

### UI accessibility concerns?

None, this is a backend/mobile refactor

## Added tests?

      have not been included_

## Added to documentation?

      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/)

## [optional] Are there any post deployment tasks we need to perform?

Double check the Pusher Beams keys ENV variables were transferred correctly to the SiteConfig (the PR defaults these values with whatever's available on `ApplicationConfig["PUSHER_BEAMS_ID"]` and `ApplicationConfig["PUSHER_BEAMS_KEY"]`).

## [optional] What gif best describes this PR or how it makes you feel?

![skating through life](https://media.giphy.com/media/l378a7b6yVbgFHIsg/giphy.gif)
